include/runtime/collections/RBTree.h

```diff
@@ -0,0 +1,512 @@
+#ifndef RBTREE_HEADER_H
+#define RBTREE_HEADER_H
+
+/* Adapted from https://github.com/BartoszMilewski/Okasaki/tree/master/RBTree */
+/* Implementing algorithms from                                               *
+ * https://matt.might.net/papers/germane2014deletion.pdf                      */
+
+#include <cassert>
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+#define CONSTRUCT_MSG_AND_THROW(msg)                                           \
+  do {                                                                         \
+    std::string emsg = std::string("[") + std::string(__func__)                \
+                       + std::string("] ") + std::string(msg);                 \
+    throw std::invalid_argument(emsg);                                         \
+  } while (false)
```

Comment:
We have pulled in `fmt` as a dependency in an [in-flight PR](https://github.com/runtimeverification/llvm-backend/pull/603); we should rewrite this code to use that new infrastructure once one or the other is merged. It's fine for now as the dependency isn't available in this PR.

Reply:
The PR mentioned is merged, so this code can be moved over to FMT.

Reply:
Using FMT in implementing the error printing macro.